zproject/backends.py

```diff
@@ -697,6 +698,12 @@ def get_or_build_user(self, username: str, ldap_user: _LDAPUser) -> Tuple[UserPr
             # An existing user, successfully authed; return it.
             return user_profile, False
 
+        if UserProfile.objects.filter(ext_auth_uid={"ldap": ldap_user.attrs[settings.AUTH_LDAP_USERNAME_ATTR][0]}, realm=self._realm).exists():
+            user_profile = UserProfile.objects.get(ext_auth_uid={"ldap": ldap_user.attrs[settings.AUTH_LDAP_USERNAME_ATTR][0]}, realm=self._realm)
```

Comment:
Hmm, I think we should be storing the entire `dn` rather than just `AUTH_LDAP_USERNAME_ATTR` value, due to uniqueness issues?

Second question - how does the `(ext_auth_uid={"ldap": ldap_user.attrs[settings.AUTH_LDAP_USERNAME_ATTR][0])` query work exactly? What happens if that json dictionary under `ext_auth_uid` also has some other entries than just ldap? Will the query find the user correctly?

Reply:
I do not think we should store the whole dn, because that would mean a user can not be moved to a different ou.

Have to test that.
As this only gets set at creation of a user I think that theoretically only one entry will exist normally.

Reply:
> I do not think we should store the whole dn, because that would mean a user can not be moved to a different ou.

We will want to add support for ldap configurations supporting having multiple domains in the near future. Storing only `uid` (or similar) attribute value will introduce a mess, because uids can repeat in other domains. Yes, moving the user to a different dn will not be supported with what I'm suggesting, but I'm not sure how to reasonably resolve that, given non-uniqueness on `uid`s.

> As this only gets set at creation of a user I think that theoretically only one entry will exist normally.

Right now yes, b ut I thoughtthe intent of the `ext_auth_uid`field is also to in the future support unique ids from other auth methods? Meaning the user might have an `"ldap"` entry in this dict as well as a `"saml"` one etc. This query shouldn't break the moment we add more things to the dict.

Reply:
https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/fields/#querying-jsonfield has more info on how to do the query

Reply:
Ok...
I can onl think of two options to solve that.
1. Have the user configure the attribute saved. That way they can configure a unique one.
2. Update the dn saved on login

Reply:
If we were to support dn changes, then updating the `dn` on login sounds right to me. Though we'd have to be careful not to introduce some sneaky vulnerabilities through such a mechanism for folks with some weird ldap configurations, with duplicate `uid`s, or emails etc. I find such scenarios tricky to reason about with regard to a `dn` change-tracking mechanism. I'm not sure if we should support such a thing at the cost of this increased complexity. @timabbott thoughts?

Reply:
I really think a changing dn should be supported, but I also think that a username attribute should already be unique.
I lean more to a config option where the admin can choose a unique property (e.g. the uid property or dn or a custom id), because we do not know which attributes might change in a ldap, but the admin should know that.
For example in my use case the dn might change frequently, but the uid does normally not.

Reply:
The query should now be correct

Reply:
Did it using the dn as suggested in https://chat.zulip.org/#narrow/stream/91-code-review/topic/.2316208.20auth.3A.20Add.20changing.20ldap.20mail/near/1025506

Reply:
https://chat.zulip.org/#narrow/stream/3-backend/topic/ext_auth_id.20create_user/near/1226818

So we are going back to having an option to configure an custom attribute